src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,104 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Timeouts;
+
+internal sealed class RequestTimeoutsMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly ICancellationTokenLinker _cancellationTokenProvider;
+    private readonly IOptionsMonitor<RequestTimeoutOptions> _options;
+
+    public RequestTimeoutsMiddleware(
+        RequestDelegate next,
+        ICancellationTokenLinker cancellationTokenProvider,
+        IOptionsMonitor<RequestTimeoutOptions> options)
+    {
+        _next = next;
+        _cancellationTokenProvider = cancellationTokenProvider;
+        _options = options;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+        var timeoutMetadata = endpoint?.Metadata.GetMetadata<RequestTimeoutAttribute>();
+        var policyMetadata = endpoint?.Metadata.GetMetadata<RequestTimeoutPolicy>();
+
+        var options = _options.CurrentValue;
+
+        if (timeoutMetadata is null && policyMetadata?.Timeout is null && options.DefaultTimeout is null)
+        {
+            return _next(context);
+        }
+
+        var disableMetadata = endpoint?.Metadata.GetMetadata<DisableRequestTimeoutAttribute>();
+        if (disableMetadata is not null)
+        {
+            return _next(context);
+        }
+
+        TimeSpan? timeSpan;
+        if (policyMetadata is not null)
+        {
+            timeSpan = policyMetadata.Timeout;
+        }
+        else if (timeoutMetadata is not null)
+        {
+            if (timeoutMetadata.Timeout is not null)
+            {
+                timeSpan = timeoutMetadata.Timeout.Value;
+            }
+            else
+            {
+                if (options.Policies.TryGetValue(timeoutMetadata.PolicyName!, out var policy))
+                {
+                    timeSpan = policy.Timeout;
+                }
+                else
+                {
+                    throw new InvalidOperationException($"The requested timeout policy '{timeoutMetadata.PolicyName}' is not available.");
+                }
+            }
+        }
+        else
+        {
+            timeSpan = options.DefaultTimeout;
+        }
+
+        if (timeSpan is null || timeSpan == Timeout.InfiniteTimeSpan)
+        {
+            return _next(context);
+        }
+
+        return SetTimeoutAsync();
+
+        async Task SetTimeoutAsync()
+        {
+            var originalToken = context.RequestAborted;
+            var linkedCts = _cancellationTokenProvider.GetLinkedCancellatinTokeSource(originalToken, timeSpan.Value);
+            try
+            {
+                context.RequestAborted = linkedCts.Token;
+                await _next(context);
+            }
+            catch (OperationCanceledException)
```

Comment:
Is it safe to assume users use `context.RequestAborted.ThrowIfCancellationRequested()` and here only catch the exception that is coming from `context.RequestAborted`? The aim is that we do not  catch other timeout exceptions that are coming from users code.
```suggestion
            catch (OperationCanceledException operationCanceledException) when (operationCanceledException.CancellationToken == context.RequestAborted)

```

Reply:
No, that would cause us to miss some exceptions such as those caused by chained tokens. The `linkedCts.Token.IsCancellationRequested && !originalToken.IsCancellationRequested are intended to avoid other cancellations.